docs/geps/gep-04.rst

```diff
@@ -0,0 +1,355 @@
+correspond to columns in the data, i.e., nodes in the graph.
+
+Unless functions perform aggregations, they are written in terms of scalars and
+vectorized during DAG setup.
+
+
+
+Motivation
+----------
+
+The implementation choice to use a DAG to represent the taxes and transfers system is
+motivated by two main reasons.
+
+1. The taxes and transfers system is constantly evolving in many dimensions,
+   flexibility is thus needed internally. Additionally, it is not enough to represent
+   the state of the system at any given point in time, but users need to be able to
+   introduce their own changes. Being able to change or replace any part of the taxes
+   and transfers system is crucial. Put differently, there is no meaningful distinction
+   between parts of the system only ever touched by developers and parts that are
+   modifiable by users. A DAG implementation allows to eliminate this usual boundary
+   for almost all use cases.
+
+1. The DAG allows a user to limit computations to generate a set of target variables,
+   which she is ultimately interested in. Doing so allows cutting down on the number of
+   input variables, it prevents unnecessary calculations, and it increases computation
+   speed.
+
+
+Basic idea
+----------
+
+Based on the two requirements above we split the taxes and transfers system into a set
+of small functions. Each function calculates one clearly defined variable (identical to
+the function's stringified name) and returns a 1d-array.
+
+.. note::
+
+    The function code itself will typically work on scalars and is vectorized by
+    GETTSIM; this is irrelevant for the DAG.
+
+Function arguments can be of three kinds:
+
+- User-provided input variables (e.g., ``bruttolohn_m``).
+- Outputs of other functions in the taxes and transfers system (e.g., ``eink_st_tu``).
+- Parameters of the taxes and transfers system, which are pre-defined and always end in
+  ``_params`` (e.g., ``ges_rentenv_params``).
+
+GETTSIM will calculate the variables a researcher is interested in by starting with the
+input variables and calling the required functions in a correct order. This is
+accomplished via a DAG (see below).
+
+Splitting complex calculations into smaller pieces has a lot of the usual advantages of
+why we use functions when programming: readability, simplicity, lower maintenance costs
+(single-responsibility principle). Another advantage is that each function is a
+potential entry point for a researcher to change the taxes and transfers system if she
+is able to replace this function with her own version.
+
+See the following example for capital income taxes.
+
+.. code-block:: python
+
+    def abgelt_st_tu(
+        zu_verst_kapital_eink_tu: FloatSeries, abgelt_st_params: dict
+    ) -> FloatSeries:
+        """Calculate abgeltungssteuer per tax unit.
+
+        Parameters
+        ----------
+        zu_verst_kapital_eink_tu
+            See :func:`zu_verst_kapital_eink_tu`.
+        abgelt_st_params
```

Comment:
I would prefer not to repeat the function name in the params argument. I.e. just `params` instead of `repeated_function_name_params`. This could either be achieved with the partial approach or with a similar approach as in my `dags` PR from yesterday. Here is a [test case](https://github.com/OpenSourceEconomics/dags/blob/2b081b9c92217db1732e81e9099436e78b038fcf/tests/test_dag.py#L124) that illustrates what I mean.

Reply:
It is actually not `repeated_function_name_params` but the first part is associated to a parameter group (e.g. `eink_st`, `wohngeld`, etc.). Just a coincidence that the parameter group is the same as the function here.

I think we need to and should keep the parameter group in the argument. (sometimes even parameters from two parameter groups are used)